File: patch.py
#!/usr/bin/python3
# patch.py 
import os,sys 

f = open(sys.argv[1], "rb")
base = f.read()
f.close()

f = open(sys.argv[2], "rb")
ips = f.read()
f.close()

_hdr = ips[:5]
if _hdr != b'PATCH':
    print("Not a valid IPS file!")
    sys.exit()

bc = 5

class patch:
    def __init__(self):
        self.offset = 0
        self.length = 0
        self.bytes = []
    ###
###
patches=[]
while ips[bc:bc+3] != b'EOF':
    p = patch()
    p.offset = (ips[bc] << 16)|(ips[bc+1] << 8)|(ips[bc+2]) 
    bc += 3
    p.length = (ips[bc] << 8)|ips[bc+1]
    bc += 2
    p.bytes = ips[bc:bc+p.length]
    bc += p.length 
    patches.append(p)

for p in patches:
    base = base[:p.offset] + p.bytes + base[p.offset+p.length:]

f = open(sys.argv[1]+".out", "wb")
f.write(base)
f.close()
print(sys.argv[1]+".out written.")

File: patch.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define u8 unsigned char


typedef struct _patch { 
    long offset;
    int size;
    u8* bytes;
} patch;

void not_ips(){
    printf("Not an IPS patch file!");
}

int main(int argc, char *argv[]) {
    if (argc < 3) {
        printf("Give 2 files as arguments");
        return 1; // Not enough arguments
    }

    FILE *f;
    f = fopen(argv[1], "rb");
    if (!f) return 1; // Error opening file
    fseek(f, 0, SEEK_END);
    long base_size = ftell(f);
    fseek(f, 0, SEEK_SET);
    unsigned char *base = (unsigned char *)malloc(base_size);
    fread(base, 1, base_size, f);
    fclose(f);

    f = fopen(argv[2], "rb");
    if (!f) {
        free(base);
        return 1; // Error opening file
    }
    fseek(f, 0, SEEK_END);
    long ips_size = ftell(f);
    fseek(f, 0, SEEK_SET);
    unsigned char *ips = (unsigned char *)malloc(ips_size);
    fread(ips, 1, ips_size, f);
    fclose(f);

    int changes_capacity = 100;
    int changes_size = 0;
    patch* changes = (patch *)malloc(changes_capacity * sizeof(patch));

    int i = 0;
    if(ips[i] != 'P') { not_ips(); return 1; }
    else if(ips[i+1] != 'A') { not_ips(); return 1; }
    else if(ips[i+2] != 'T') { not_ips(); return 1; }
    else if(ips[i+3] != 'C') { not_ips(); return 1; }
    else if(ips[i+4] != 'H') { not_ips(); return 1; }
    
    for(i = 5; i < ips_size; i++) {
        if(ips[i] == 'E')
            if(ips[i+1] == 'O')
                if(ips[i+2] == 'F')
                    break;
        changes[changes_size].offset = (ips[i] << 16) | (ips[i+1] << 8) | (ips[i+2]);
        i += 3;
        changes[changes_size].size = (ips[i]<<8)|ips[i+1];
        i += 2;
        changes[changes_size].bytes = (u8*)malloc(sizeof(u8) * changes[changes_size].size);
        for(int j = 0; j < changes[changes_size].size; j++){
            changes[changes_size].bytes[j] = ips[i];
            i++;
        }
        i--;
        printf("%lu %d\n", changes[changes_size].offset, changes[changes_size].size);
        changes_size++;
        if(changes_size >= changes_capacity){
            changes_capacity += 100;
            changes = (patch *)realloc(changes, changes_capacity * sizeof(patch));
        }
        
    }
    // now adjust input bytes 
    for(i = 0; i < changes_size; i++){
        printf("a");
        int p = 0;
        for(int j = changes[i].offset; j < changes[i].offset + changes[i].size; j++){
            base[j] = changes[i].bytes[p++];
        }
    }
        
    f = fopen("out.bin", "wb");
    fwrite(base, 1, base_size, f);
    fclose(f);
    printf("out.bin file written\n");

    free(base);
    free(ips);
    for(int i = 0; i < changes_size; i++){
        free(changes[i].bytes);
    }
    free(changes);

    return 0;
}
